### PR TITLE
Refactor topology discovery service tests

### PR DESCRIPTION
**Pull request by andrewkerr9000**
Refactor some of the testing infrastructure around DiscoveryService, primarily introduce an enum of DiscoveryServiceFactory with constructors.

### CODE REVIEW

File: enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/scenarios/ClusterOverviewIT.java [hunk 1]
```diff
@@ -171,20 +158,22 @@ public void shouldDiscoverReadReplicasAfterRestartingCores() throws Exception
...
         clusterRule.withNumberOfReadReplicas( 0 );
 
         Cluster cluster = clusterRule.startCluster();
 
         // when
-        cluster.addCoreMemberWithId( 3 ).start();
-        cluster.addCoreMemberWithId( 4 ).start();
+        int extraCoreMembers = 2;
+        int finalCoreMembers = initialCoreMembers + 2;
+        IntStream.range( 0, extraCoreMembers ).forEach( idx -> cluster.addCoreMemberWithId( initialCoreMembers + idx ).start() );
```
Comment (hugofirth) [inline enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/scenarios/ClusterOverviewIT.java]: If we're going to go this way: "documenting" with variable names, should we factor out / replicate to make it consistent across the other tests in this IT?

**Review by hugofirth — APPROVED** [review]
One query about the ClusterOverviewIT, but otherwise looks good to me 👍

File: enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/scenarios/ClusterOverviewIT.java (hunk 1 above)
Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/scenarios/ClusterOverviewIT.java]: Baby steps... I'll update the other tests just as soon as I care about them :)

Comment (andrewkerr9000) [reply enterprise/causal-clustering/src/test/java/org/neo4j/causalclustering/scenarios/ClusterOverviewIT.java]: although the `2` on line 169 looks like it should be `extraCoreMembers`...

**Pull request merged**